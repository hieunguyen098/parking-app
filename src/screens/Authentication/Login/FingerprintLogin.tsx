import { StyleSheet, Text, View, Pressable, Switch, Alert } from 'react-native';
import React, { useEffect, useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import * as LocalAuthentication from 'expo-local-authentication';
import * as SecureStore from 'expo-secure-store';
import { FontAwesome5 } from '@expo/vector-icons';
import styles from '../styles';
import { login } from '../../../services/auth';
import { useSelector } from 'react-redux';
import AsyncStorage from '@react-native-async-storage/async-storage';
function FingerprintLogin({ handleFingerprint, setIsLoading }: any) {
    const [isBiometricSupported, setIsBiometricSupported] = useState(false);
    const user = useSelector((state: any) => state.auth.user);
    const navigation: any = useNavigation();
    useEffect(() => {
        (async () => {
            const compatible = await LocalAuthentication.hasHardwareAsync();
            setIsBiometricSupported(compatible);
        })();
    });

    const fallBackToDefaultAuth = () => {
        console.log('fall back to password authentication');
    };

    const alertComponent = (title: any, mess: any, btnTxt: any, btnFunc: any) => {
        return Alert.alert(title, mess, [
            {
                text: btnTxt,
                onPress: btnFunc,
            },
        ]);
    };

    const handleLogin = async () => {
        setIsLoading(true);
        const password = await SecureStore.getItemAsync('password');
        console.log(password, user.phone);
        const response = await login(user.phone, password as string);
        console.log(response);
        if (response.returnCode > 0) {
            navigation.navigate('TabBarScreen');
            if (!response.data) {
                return;
            }
            if (response.data[0].needOtp) {
                navigation.navigate('PhoneVerification');
            } else {
                // AsyncStorage.setItem('accessToken', response.data[0].accessToken);
            }
        }
        setIsLoading(false);
    };

    const handleBiometricAuth = async () => {
        const isBiometricAvailable = await LocalAuthentication.hasHardwareAsync();
        if (!isBiometricAvailable)
            return alertComponent('Please enter your password', 'Biometric Authentication not supported', 'OK', () =>
                fallBackToDefaultAuth(),
            );

        // Check Biometrics types available (Fingerprint, Facial recognition, Iris recognition)
        let supportedBiometrics;
        if (isBiometricAvailable) supportedBiometrics = await LocalAuthentication.supportedAuthenticationTypesAsync();

        // Check Biometrics are saved locally in user's device
        const savedBiometrics = await LocalAuthentication.isEnrolledAsync();
        if (!savedBiometrics)
            return alertComponent('Biometric record not found', 'Please login with your password', 'OK', () =>
                fallBackToDefaultAuth(),
            );

        // Authenticate use with Biometrics (Fingerprint, Facial recognition, Iris recognition)

        const biometricAuth = await LocalAuthentication.authenticateAsync({
            promptMessage: 'Login with Biometrics',
            cancelLabel: 'Cancel',
            disableDeviceFallback: true,
        });

        if (biometricAuth.success) {
            handleLogin();
            console.log('success');
        } else if (biometricAuth.error) console.log('error', biometricAuth.error);

        console.log({ isBiometricAvailable });
        console.log({ supportedBiometrics });
        console.log({ savedBiometrics });
        console.log({ biometricAuth });
    };
    return (
        <Pressable onPress={handleBiometricAuth} style={styles.fingerprintContainer}>
            <FontAwesome5 name="fingerprint" size={24} color="black" />
            <Text style={styles.fingerprintBtn}>Mở khóa bằng vân tay</Text>
        </Pressable>
    );
}

export default FingerprintLogin;
