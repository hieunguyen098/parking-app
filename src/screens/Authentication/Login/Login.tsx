import { Pressable, StyleSheet, Text, View } from 'react-native';
import React, { useState } from 'react';
import PasswordInput from '../../../components/PasswordInput/PasswordInput';
import LargeButton from '../../../components/Buttons/LargeButton';
import { useNavigation } from '@react-navigation/native';
import styles from '../styles';
import { useSelector } from 'react-redux';
import { login } from '../../../services/auth';

const Login = () => {
    const navigation: any = useNavigation();
    const user = useSelector((state: any) => state.auth.user);
    const [pin, setPin] = useState('');
    const navigate = (name: string) => {
        navigation.navigate(name, { nextScreen: 'TabBarScreen', action: 'login' });
    };
    const onContinue = async () => {
        const res = await login(user.phone, pin);
        console.log(user)
        console.log(res);
    };
    return (
        <View style={styles.container}>
            <View style={styles.group}>
                <Text style={styles.description}>Xin chào, {user.fullname}!</Text>
                <Text style={styles.description}>{user.phone}</Text>
                <PasswordInput length={6} value={pin} setValue={setPin} />
                <Pressable onPress={() => navigate('ForgotPassword')}>
                    <Text style={styles.forgotPasswordBtn}>Quên mật khẩu?</Text>
                </Pressable>
            </View>
            <LargeButton onPress={onContinue} title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default Login;
