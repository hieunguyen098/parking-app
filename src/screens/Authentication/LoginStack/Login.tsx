import { Pressable, StyleSheet, Text, View } from 'react-native';
import React from 'react';
import PasswordInput from '../components/PasswordInput/PasswordInput';
import LargeButton from '../../../components/Buttons/LargeButton';
import { useNavigation } from '@react-navigation/native';
import styles from '../styles';

const Login = () => {
    const navigation: any = useNavigation();
    const navigate = (name: string) => {
        navigation.navigate(name);
    };
    return (
        <View style={styles.container}>
            <View style={styles.group}>
                <Text style={styles.description}>Xin chào, CAO THANH BÌNH!</Text>
                <Text style={styles.description}>0398888888</Text>
                <PasswordInput length={6} />
                <Pressable onPress={() => navigate('ForgotPasswordStack')}>
                    <Text style={styles.forgotPasswordBtn}>Quên mật khẩu?</Text>
                </Pressable>
            </View>
            <LargeButton onPress={() => navigate('PhoneVerification')} title="Tiếp tục" style={styles.continueButton} />
        </View>
    );
};

export default Login;
