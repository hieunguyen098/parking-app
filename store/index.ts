import { configureStore } from '@reduxjs/toolkit';
import { authReducer } from './slices/authSlice';
import { generalReducer } from './slices/generalSlice';

export const store = configureStore({
    reducer: {
        auth: authReducer,
        general: generalReducer,
    },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
