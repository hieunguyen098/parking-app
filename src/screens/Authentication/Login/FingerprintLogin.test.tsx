import React from 'react';
import { render } from '@testing-library/react-native';
import { NavigationContainer } from '@react-navigation/native';

import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { store } from '../../../../store';
import Login from './Login';
import FingerprintLogin from './FingerprintLogin';

jest.useFakeTimers();
test('test login', () => {
    const queryClient = new QueryClient();
    const component = (
        <QueryClientProvider client={queryClient}>
            <Provider store={store}>
                <NavigationContainer>
                    <FingerprintLogin />
                </NavigationContainer>
            </Provider>
        </QueryClientProvider>
    );

    const tree = render(component);
    expect(tree).toMatchSnapshot();
});
