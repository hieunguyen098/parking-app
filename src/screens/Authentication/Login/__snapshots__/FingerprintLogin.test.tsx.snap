// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test login 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
      "marginTop": 16,
    }
  }
>
  <Text />
  <Text
    style={
      Object {
        "fontSize": 16,
        "paddingLeft": 8,
      }
    }
  >
    Mở khóa bằng vân tay
  </Text>
</View>
`;
